### PR TITLE
MGMT-7090: Bind and Unbind Agent

### PR DESCRIPTION
**Pull request by rollandf**
Signed-off-by: Fred Rolland <frolland@redhat.com>

# Assisted Pull Request

## List all the issues related to this PR

- [ ] New Feature
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [x] No tests needed

## Assignees

/cc @filanov 
/cc @yevgeny-shnaidman 
/cc @nmagnezi 

## Checklist

- [ ] Title and description added to both, commit and PR.
- [ ] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [ ] Reviewers have been listed
- [ ] This change does not require a documentation update (docstring, `docs`, README, etc)
- [ ] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: internal/controller/controllers/agent_controller.go [hunk 1]
```diff
@@ -127,68 +123,71 @@ func (r *AgentReconciler) Reconcile(origCtx context.Context, req ctrl.Request) (
...
-	if err != nil {
-		if errors.Is(err, gorm.ErrRecordNotFound) {
-			// Delete the agent, using a finalizer with pre-delete to deregister the host.
-			log.Infof("Cluster name: %s namespace: %s not found in backend, deleting Agent",
-				agent.Spec.ClusterDeploymentName.Name, agent.Spec.ClusterDeploymentName.Namespace)
-			return r.deleteAgent(ctx, log, req.NamespacedName)
-		}
-		// Update that we failed to retrieve the cluster from the database
-		return r.updateStatus(ctx, log, agent, nil, nil, err, !errors.Is(err, gorm.ErrRecordNotFound))
+	if agent.Spec.ClusterDeploymentName != nil && h.ClusterID == nil{
```
Comment (yevgeny-shnaidman) [inline internal/controller/controllers/agent_controller.go:145]: i think you can move it under the if agent.Spec.ClusterDeploymentName != nil at lne 150, since you will need to the clusterdeployment anyway, no?

File: api/v1beta1/infraenv_types.go [hunk 2]
```diff
@@ -60,7 +60,6 @@ type InfraEnvSpec struct {
 	// +optional
 	SSHAuthorizedKey string `json:"sshAuthorizedKey,omitempty"`
 
-	// This field is currently not used. The Pull Secret from ClusterDeployment will be used for the ISO creation.
```
Comment (filanov) [inline api/v1beta1/infraenv_types.go:63]: if you are removing this part you need to remove the update from cluster deployment controller https://github.com/openshift/assisted-service/blob/8eedf068bcc104c66a1334a94e54598aa507b9ee/internal/controller/controllers/clusterdeployments_controller.go#L716-L720
Suggesting to split it into a different PR.

Comment (rollandf) [reply api/v1beta1/infraenv_types.go]: Sure

File: internal/controller/controllers/agent_controller.go [hunk 3]
```diff
@@ -127,68 +123,98 @@ func (r *AgentReconciler) Reconcile(origCtx context.Context, req ctrl.Request) (
...
+		return r.updateStatus(ctx, log, agent, &host.Host, h.ClusterID, nil, true)
 	}
 
-	// Retrieve cluster by ClusterDeploymentName from the database
-	cluster, err := r.Installer.GetClusterByKubeKey(kubeKey)
-	if err != nil {
-		if errors.Is(err, gorm.ErrRecordNotFound) {
-			// Delete the agent, using a finalizer with pre-delete to deregister the host.
-			log.Infof("Cluster name: %s namespace: %s not found in backend, deleting Agent",
+	if agent.Spec.ClusterDeploymentName != nil {
```
Comment (filanov) [inline internal/controller/controllers/agent_controller.go:148]: We need to handle a case when user just replace the cluster deployment reference.
So we will need to unbind and then bind to the new cluster.

Comment (rollandf) [reply internal/controller/controllers/agent_controller.go:148]: Sure, do you want to handle this in this PR or on a follow up?

Comment (filanov) [reply internal/controller/controllers/agent_controller.go:148]: I don't mind to split it. just making sure that we don't forget about it because backend will not handle it

Comment (rollandf) [reply internal/controller/controllers/agent_controller.go:148]: I opened https://issues.redhat.com/browse/MGMT-7732 for this scenario.

**Review by filanov — APPROVED** [review]
